src/tracers/tracers.cpp

```diff
@@ -0,0 +1,293 @@
+
+            // TODO(pgrete) check if this actually required
+            bool on_current_mesh_block = true;
+            swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), on_current_mesh_block);
+          });
+    }
+  } else {
+    PARTHENON_THROW("Unknown tracer initial_seed_method");
+  }
+}
+
+TaskStatus AdvectTracers(MeshBlockData<Real> *mbd, const Real dt) {
+  auto *pmb = mbd->GetParentPointer();
+  auto &sd = pmb->meshblock_data.Get()->GetSwarmData();
+  auto &swarm = sd->Get("tracers");
+
+  auto &x = swarm->Get<Real>(swarm_position::x::name()).Get();
+  auto &y = swarm->Get<Real>(swarm_position::y::name()).Get();
+  auto &z = swarm->Get<Real>(swarm_position::z::name()).Get();
+  auto &vel_x = swarm->Get<Real>("vel_x").Get();
+  auto &vel_y = swarm->Get<Real>("vel_y").Get();
+  auto &vel_z = swarm->Get<Real>("vel_z").Get();
+
+  auto swarm_d = swarm->GetDeviceContext();
+
+  const auto &prim_pack = mbd->PackVariables(std::vector<std::string>{"prim"});
+
+  // update loop. RK2
+  const int max_active_index = swarm->GetMaxActiveIndex();
+  pmb->par_for(
+      "Advect Tracers", 0, max_active_index, KOKKOS_LAMBDA(const int n) {
+        if (swarm_d.IsActive(n)) {
+          int k, j, i;
+          swarm_d.Xtoijk(x(n), y(n), z(n), i, j, k);
+
+          // RK2/Heun's method (as the default in Flash)
+          // https://flash.rochester.edu/site/flashcode/user_support/flash4_ug_4p62/node130.html#SECTION06813000000000000000
+          // Intermediate position and velocities
+          // x^{*,n+1} = x^n + dt * v^n
+          const auto x_star = x(n) + dt * vel_x(n);
+          const auto y_star = y(n) + dt * vel_y(n);
+          const auto z_star = z(n) + dt * vel_z(n);
+
+          // v^{*,n+1} = v(x^{*,n+1}, t^{n+1})
+          // First parameter b=0 assume to operate on a pack of a single block and needs
+          // to be updated if this becomes a MeshData function
+          const auto vel_x_star = LCInterp::Do(0, x_star, y_star, z_star, prim_pack, IV1);
+          const auto vel_y_star = LCInterp::Do(0, x_star, y_star, z_star, prim_pack, IV2);
+          const auto vel_z_star = LCInterp::Do(0, x_star, y_star, z_star, prim_pack, IV3);
+
+          // Full update using mean velocity
+          x(n) += dt * 0.5 * (vel_x(n) + vel_x_star);
+          y(n) += dt * 0.5 * (vel_y(n) + vel_y_star);
+          z(n) += dt * 0.5 * (vel_z(n) + vel_z_star);
+
+          // The following call is required as it updates the internal block id
+          // following the advection. The internal id is used in the subsequent task to
+          // communicate particles.
+          bool unused_temp = true;
+          swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), unused_temp);
+        }
+      });
+
+  return TaskStatus::complete;
+} // AdvectTracers
+
+/**
+ * FillDerived function for tracers.
+ * Registered Quantities (in addition to t, x, y, z):
+ * rho, vel, B
+ **/
+TaskStatus FillTracers(MeshData<Real> *md, parthenon::SimTime &tm) {
```

Comment:
Is this called before or after the timestep? If after, how are the initial particle velocities set?

Reply:
Good point. So far it's called at the end of the last stage of each timestep so initial values are all 0.
I now added another call to the end of the seeding routine so that the data is initialized properly in cases where the initial data is non-zero.